### PR TITLE
fix(artifacts): Consider artifacts from stages when triggering from a pipeline stage

### PR DESCRIPTION
**Pull request by sergio-quintero**
Follow up to https://github.com/spinnaker/orca/pull/4489

Triggering a pipeline from Pipeline A using the [pipeline stage](https://spinnaker.io/docs/reference/pipeline/stages/#pipeline) where Pipeline B uses BakeManifestStage or DeployManifestStage which rely on expectedArtifacts is broken. [ArtifactUtils](https://github.com/spinnaker/orca/blob/master/orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactUtils.java#L264-L275) is filtering the expectedArtifacts needed in stages. 

In 1.28, you were able to define expectedArtifacts in Pipeline B, and when you triggered from Pipeline A, Pipeline B was able to resolve the artifacts from Pipeline A. [Passing artifacts between pipelines](https://spinnaker.io/docs/reference/ref-artifacts/in-pipelines/#passing-artifacts-between-pipelines)

### CODE REVIEW

**Review by jervi — APPROVED** [review]
Looks good to me! Tagging @xibz as well, since you also showed some interest in this area

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy [hunk 1]
```diff
@@ -85,12 +85,24 @@ class DependentPipelineStarter implements ApplicationContextAware {
...
+    // following a similar approach as triggers above
     // expectedArtifacts can be used in triggers and stages
-    // for now we identified DeployManifestStage
-    // in ResolveDeploySourceManifestTask using ManifestEvaluator.getRequiredArtifacts
-    def requiredArtifactIds = pipelineConfig.get("stages", []).collectMany {
-      it.requiredArtifactIds ?: []
+    def requiredArtifactIds = pipelineConfig.get("stages", []).collectMany { stage ->
+      def artifactIds = []
+
+      //DeployManifestStage in ResolveDeploySourceManifestTask using ManifestEvaluator.getManifestArtifact
```
Comment (xibz) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:93]:
```suggestion
      // DeployManifestStage in ResolveDeploySourceManifestTask using ManifestEvaluator.getManifestArtifact
```
nit

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy [hunk 2]
```diff
@@ -85,12 +85,24 @@ class DependentPipelineStarter implements ApplicationContextAware {
...
+      artifactIds += stage.manifestArtifactId ?: []
+
+      // DeployManifestStage in ResolveDeploySourceManifestTask using ManifestEvaluator.getRequiredArtifacts
+      artifactIds += stage.requiredArtifactIds ?: []
+
+      // BakeManifestStage in CreateBakeManifestTask using BakeManifestContext for Kustomize
+      artifactIds += stage.inputArtifact?.id ?: []
+
+      // BakeManifestStage in CreateBakeManifestTask using BakeManifestContext for Helm
```
Comment (xibz) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:102]: Are these comments accurate? What I mean is I understand that `Helm` bakes with `inputArtifacts` but other stages like `BakeCloudFoundry` also utilize this?

Can we make this more generic and say something more like

```
// Some bake manifest stages may populate the inputArtifacts array. So we need to look to for those ids as well
```

And we may be able to get rid of the other comments as they seem to be very specific but don't capture the whole picture

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy [hunk 3]
```diff
@@ -85,12 +85,24 @@ class DependentPipelineStarter implements ApplicationContextAware {
...
     }
 
-    // we are following a similar approach as triggers above
+    // following a similar approach as triggers above
     // expectedArtifacts can be used in triggers and stages
-    // for now we identified DeployManifestStage
-    // in ResolveDeploySourceManifestTask using ManifestEvaluator.getRequiredArtifacts
-    def requiredArtifactIds = pipelineConfig.get("stages", []).collectMany {
-      it.requiredArtifactIds ?: []
+    def requiredArtifactIds = pipelineConfig.get("stages", []).collectMany { stage ->
```
Comment (xibz) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:90]: Do we need this variable? Can't we just use `expectedArtifactIds` here?

**Review by xibz — CHANGES_REQUESTED** [review]
Looks good! I have some very minor nit comments

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy (hunk 2 above)
Comment (xibz) [reply orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:102]: Maybe even explain why we are looking for these ids as opposed to the expected artifact. Would be good to just document here saying we dont know how consuming pipelines will utilize artifacts from the parent pipeline, so we rely on any input artifact as opposed to the expected artifact

**Review by edgarulg — APPROVED** [review]
LGTM! but before merging it let's address all comments :)

**Comment by sergio-quintero** [comment]
Closing this PR because we found more stages with different field names that were not considered in this PR. Adding a list of fields in `DependentPipelineStarter` is not maintainable. Also, if someone creates a plugin with a new Stage that utilizes expectedArtifacts, they can define any field name which will not be included in this class. 

We'll send a new PR with a better fix for this issue.

**Pull request closed without merging**